tests/test-cast-to-overflow.c

```diff
@@ -0,0 +1,91 @@
+
+/* Unit test trucate in type cast.
+   Copyright (C) 2016 Free Software Foundation, Inc.
+   This file is part of the Decimal Floating Point C Library.
+   The Decimal Floating Point C Library is free software; you can
+   redistribute it and/or modify it under the terms of the GNU Lesser
+   General Public License version 2.1.
+   The Decimal Floating Point C Library is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+   the GNU Lesser General Public License version 2.1 for more details.
+   You should have received a copy of the GNU Lesser General Public
+   License version 2.1 along with the Decimal Floating Point C Library;
+   if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+   Suite 330, Boston, MA 02111-1307 USA.
+   Please see libdfp/COPYING.txt for more information.  */
+
+
+#ifndef __STDC_WANT_DEC_FP__
+#define __STDC_WANT_DEC_FP__
+#endif
+
+#include <float.h>
+#include <stdio.h>
+#include <fenv.h>
+#include <math.h>
+
+#define _WANT_VC 1
+#include "scaffold.c"
+
+typedef struct
+{
+  int line;
+  int round_mode;
+  _Decimal128 set;
+  double expect;
+} d_type;
+
+/* Decimal 128 to Binary Double 64.  */
+static const d_type d128to64[] = {
+ {__LINE__, FE_TONEAREST, 1e1000DL, INFINITY},
+ {__LINE__, FE_TONEAREST, -1e1000DL, -INFINITY},
+ {__LINE__, FE_TOWARDZERO, 1e1000DL, DBL_MAX},
+ {__LINE__, FE_TOWARDZERO, -1e1000DL, -DBL_MAX},
+ {__LINE__, FE_DOWNWARD, 1e1000DL, DBL_MAX},
+ {__LINE__, FE_DOWNWARD, -1e1000DL, -INFINITY},
+ {__LINE__, FE_UPWARD, 1e1000DL, INFINITY},
+ {__LINE__, FE_UPWARD, -1e1000DL, -DBL_MAX}
+};
+
+/* Decimal 128 to Binary Float 32.  */
+static const d_type d128to32[] = {
+ {__LINE__, FE_TONEAREST, 1e1000DL, INFINITY},
+ {__LINE__, FE_TONEAREST, -1e1000DL, -INFINITY},
+ {__LINE__, FE_TOWARDZERO, 1e1000DL, FLT_MAX},
+ {__LINE__, FE_TOWARDZERO, -1e1000DL, -FLT_MAX},
+ {__LINE__, FE_DOWNWARD, 1e1000DL, FLT_MAX},
+ {__LINE__, FE_DOWNWARD, -1e1000DL, -INFINITY},
+ {__LINE__, FE_UPWARD, 1e1000DL, INFINITY},
+ {__LINE__, FE_UPWARD, -1e1000DL, -FLT_MAX}
+};
+
+static const int d128to64_size = sizeof (d128to64) / sizeof (d128to64[0]);
+
+static const int d128to32_size = sizeof (d128to32) / sizeof (d128to32[0]);
+
+int
+main(void)
+{
+  int i;
+  for (i=0; i<d128to64_size; ++i)
+    {
+      double ret;
+      fesetround (d128to64[i].round_mode);
+      ret = (double) d128to64[i].set;
+      _VC_P (__FILE__, d128to64[i].line, d128to64[i].expect, ret, "%a");
+    }
+
+  for (i=0; i<d128to32_size; ++i)
+    {
+      float ret;
+      fesetround (d128to32[i].round_mode);
+      ret = (float) d128to32[i].set;
+      _VC_P (__FILE__, d128to32[i].line, d128to32[i].expect, ret, "%a");
+    }
+
```

Comment:
Do we want to test for long double, too?

Reply:
Isn't long double and double the same on POWER?

Reply:
No.  It might have been in the past, but it is not anymore.  On POWER, long double uses two 64-bits registers to get more precision (without changing the range, i.e.: no changes in the exponent size).

https://en.wikipedia.org/wiki/IBM_Floating_Point_Architecture#Extended-precision_128-bit

Reply:
Turns out we found a new bug. Conversion to long double returns a NaN instead Infinity. I've opened a new task (#60) to fix this problem. It's a different kind of problem since on that case it's not related to the rounding mode but to the conversion itself. (It's something before the rounding mode switch).

Reply:
OK.  Thanks for the new bug report.
You don't need to address that in this pull request.